Create new version

jflora91 opened the pull request:
create class Configuration & Timestamp(will be used by Configuration and Base class in the future)
createNewVersion method in Service
createNewVersion method in Controller
validation of tenant
creation of version number
logs and exceptions

tsbatista reviewed (commented):
Did you properly rebase this PR against the upstream master? You need to ensure you rebase your own work before creating a PR or the history will be nearly impossible to understand. I say this because I was expecting no changes to the gitignore, some changes to the pom (not new files) and a few new java files. I will only do a proper review of this once the PR diff seems a bit healthier.

ncarapeto commented on configuration-service/src/main/java/tech/plinth/config/service/ConfigurationService.java:
```diff
@@ -0,0 +1,86 @@
+            logger.error("Tenant need to be specified, field is null");
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Tenant need to be specified, field is null");
+        }
+        if (tenant.isBlank() || tenant.isEmpty()) {
```
tenant.isBlank() already tests for isEmpty

jflora91 replied:
> tenant.isBlank() already tests for isEmpty

Solved in the new commit

tsbatista commented on configuration-service/src/main/java/tech/plinth/config/Bootstrap.java:
```diff
@@ -1,11 +1,13 @@
 
 @SpringBootApplication
-public class Bootstrap
+@EnableJpaAuditing
```
What requirement prompted you to enable this?

jflora91 replied:
> What requirement prompted you to enable this?

I'm didn't implement auditing in my code, no need this annotation

tsbatista commented on configuration-service/src/main/java/tech/plinth/config/model/Timestamp.java:
```diff
@@ -0,0 +1,27 @@
+import java.time.LocalDateTime;
+
+@MappedSuperclass
+public class Timestamp {
```
Wait.... What? Why? I really have no idea why you did this! Can you please elaborate on the reasoning behind this file?

jflora91 replied:
> Wait.... What? Why? I really have no idea why you did this! Can you please elaborate on the reasoning behind this file?

I created this superclass because in the future implementation of Base class i will have the same field created_at and to don't repeat myself i will reuse this class for Base too

daveseco7 replied:
What ? i honestly did not understand the use case

tsbatista commented on configuration-service/src/main/resources/application.properties:
```diff
@@ -0,0 +1,2 @@
+spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.MySQL8Dialect
+spring.jpa.hibernate.ddl-auto = create
```
This can not be here, this setting is dangerous for your data, use flyway instead

tsbatista commented on configuration-service/src/main/resources/application-dev.properties:
```diff
@@ -0,0 +1,3 @@
+spring.datasource.url=jdbc:mysql://localhost:3306/configuration?serverTimezone=UTC
+spring.datasource.username=root
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
```
This is not needed, hibernate can auto detect this

jflora91 replied:
> This is not needed, hibernate can auto detect this

Solved

tsbatista commented on configuration-service/src/main/resources/application.properties:
```diff
@@ -0,0 +1,2 @@
+spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.MySQL8Dialect
```
This can not be here as well, let it be auto detected

jflora91 replied:
> This can not be here as well, let it be auto detected

Done

tsbatista commented on configuration-service/src/main/java/tech/plinth/config/service/ConfigurationService.java:
```diff
@@ -0,0 +1,86 @@
+
+        Configuration configurationSaved = configurationRepository.save(configuration);
+
+        logger.info("New version of configuration service created: id: {}, version: {}, tenant: {}, data: {}"
```
What code style is this?

jflora91 replied:
> What code style is this?

you are right i refactor that, check if it's ok now

tsbatista commented on configuration-service/src/main/java/tech/plinth/config/service/ConfigurationService.java:
```diff
@@ -0,0 +1,86 @@
+     * @return
+     */
+    public Long calculateNextVersionNumber() {
+        List<Configuration> allVersions = configurationRepository.findAll();
```
This will really really need to change! but lets do it at a later stage!

tsbatista commented:
@jflora91 when I ask you to rebase your code I do it with good reason... In this PR you are now attempting to delete three files that were there for good reason and should not be touched by you:
the CODEOWNERS defines the default reviewers, the Jenkinsfile makes the project build on our Jenkins (I will show it to you one day) and the LICENSE is there for legal reasons. Please do not touch any of these files.

jflora91 commented:
> @jflora91 when I ask you to rebase your code I do it with good reason... In this PR you are now attempting to delete three files that were there for good reason and should not be touched by you:
> the CODEOWNERS defines the default reviewers, the Jenkinsfile makes the project build on our Jenkins (I will show it to you one day) and the LICENSE is there for legal reasons. Please do not touch any of these files.

I'm sorry, (i know, i made this mistake again).
For some reason i collect this files to my branch. I will not make this mistake again.

tsbatista commented on configuration-service/src/main/resources/db/migration/V1__create_initial_schema.sql:
```diff
@@ -0,0 +1,8 @@
+    tenant varchar(255) not null,
+    version bigint not null,
+    primary key (id)
+) engine=InnoDB AUTO_INCREMENT=1;
```
after the `)`, remove everything. As is, this will only run on mysql, meaning it should be in a vendor specific package. As this create table is very simple there is no reason to lock in

jflora91 replied:
> after the `)`, remove everything. As is, this will only run on mysql, meaning it should be in a vendor specific package. As this create table is very simple there is no reason to lock in

I understand your point, it was removed.

jflora91 committed:
Create new Version
create class Configuration
create package (controller, database,delegate,interceptor)
implement flyway (migration)
update pom files

jflora91 committed:
update schema migration

tsbatista commented on configuration-service/src/main/java/tech/plinth/config/delegate/ConfigurationDelegate.java:
```diff
@@ -0,0 +1,64 @@
+                    .max(Long::compare).get();
+        }
+
+//        logger.info("The last configuration created as the version: {}", version);
```
If you want to do this, it is a good place  to log at the debug or trace level

tsbatista commented on configuration-service/src/main/java/tech/plinth/config/delegate/ConfigurationDelegate.java:
```diff
@@ -0,0 +1,64 @@
+
+        Configuration configurationSaved = configurationRepository.save(configuration);
+
+//        logger.info("New version of configuration service created: id: {}, version: {}, tenant: {}, data: {}",
```
Again, if you want this, then log at a debug level, or completely remove the code

tsbatista reviewed (commented):
Overall, not it is a lot easier to read... I still need to do a deep review but I do not have the time right now

jflora91 committed:
refactor logger

daveseco7 commented on configuration-service/src/main/java/tech/plinth/config/controller/ConfigurationController.java:
```diff
@@ -0,0 +1,30 @@
+     * @return
+     */
+    @PostMapping("/config")
+    public JsonNode createNewVersion(@RequestBody JsonNode data_json) {
```
dataJson

jflora91 replied:
> dataJson

Fixed

jflora91 committed:
rename data_json to dataJson
rename tenant to platform

ncarapeto commented on configuration-service/src/main/java/tech/plinth/config/delegate/ConfigurationDelegate.java:
```diff
@@ -0,0 +1,72 @@
+     */
+    public Long calculateNextVersionNumber(String platformId) {
+
+        List<Configuration> allVersions = configurationRepository.findByPlatform(platformId);
```
Are you really going to retrieve the full table from the database? Just to find one specific line? What is the point of database queries? Just imagine you had 1 million records, would you bring them all to memory?

Test something like: 
`Optional<Configuration> findTopByPlatformOrderByVersionDesc(String platformId)`

Plus ideally all should be in a transaction (for concurrency purposes), but we can save that for later.

daveseco7 replied:
This is the way to go! If you need help achiving this ping us 😄

jflora91 replied:
> Are you really going to retrieve the full table from the database? Just to find one specific line? What is the point of database queries? Just imagine you had 1 million records, would you bring them all to memory?
> 
> Test something like:
> `Optional<Configuration> findTopByPlatformOrderByVersionDesc(String platformId)`
> 
> Plus ideally all should be in a transaction (for concurrency purposes), but we can save that for later.

Yes, you are right! I made a new implementation taking into account what you said, thank you I think it's much better now

ncarapeto commented on configuration-service/src/main/resources/db/migration/V1__create_initial_schema.sql:
```diff
@@ -0,0 +1,7 @@
+    data_json json not null,
+    platform varchar(255) not null,
+    version bigint not null,
+    primary key (id));
```
since we have nothing to prevent concurrency issues, I would add a unique constraint, just to ensure that nobody inserts two lines with the same version at the same time:
`UNIQUE(platform,version)`

daveseco7 replied:
I think that PK is always unique, even if you don't explicitly define it as UNIQUE.

jflora91 replied:
> since we have nothing to prevent concurrency issues, I would add a unique constraint, just to ensure that nobody inserts two lines with the same version at the same time:
> `UNIQUE(platform,version)`

I understand your point but if the version is calculated automatically (taking into account the latest version in DB) should we prevent something that not going to happen?

tsbatista replied:
Actually BD constraints are never too much! The if you do not add a constraint and enforce it only in code, you open your database to a problem:
Imagine you are doing maintenance on the service, manually inserting a record on a productive system because... reasons. You make the query to find out the highest version, and as you prepare the query to insert the new version someone invokes the api. You run your insert... and end up with two records containing the same version due to a race condition between you and the running service.

What would happen if the query you run to get ONE instance of the highest version returns two or more records? Try it on your local system and imagine the impact on a productive system!

jflora91 replied:
> Actually BD constraints are never too much! The if you do not add a constraint and enforce it only in code, you open your database to a problem:
> Imagine you are doing maintenance on the service, manually inserting a record on a productive system because... reasons. You make the query to find out the highest version, and as you prepare the query to insert the new version someone invokes the api. You run your insert... and end up with two records containing the same version due to a race condition between you and the running service.
> 
> What would happen if the query you run to get ONE instance of the highest version returns two or more records? Try it on your local system and imagine the impact on a productive system!

completely agree!
I have to protect my service and DB for all eventualities that may occur

ncarapeto commented on pom.xml:
```diff
@@ -30,8 +30,24 @@
+            </dependency>
+            <dependency>
+                <groupId>org.flywaydb</groupId>
+                <artifactId>flyway-core</artifactId>
```
already defined in spring-boot-dependencies (with the same version)

jflora91 replied:
> already defined in spring-boot-dependencies (with the same version)

Good point, and should i not redefine the version of mysql-connector-java because is already defined there too?

jflora91 replied:
forget, i saw now your comment about that

ncarapeto commented on pom.xml:
```diff
@@ -30,8 +30,24 @@
+            </dependency>
+            <dependency>
+                <groupId>mysql</groupId>
+                <artifactId>mysql-connector-java</artifactId>
```
not a big deal but this is already defined in spring-boot-dependencies (with a later version even)

jflora91 replied:
> not a big deal but this is already defined in spring-boot-dependencies (with a later version even)

Done

ncarapeto commented on configuration-service/src/main/resources/application.properties:
```diff
@@ -0,0 +1,5 @@
+spring.datasource.url=jdbc:mysql://localhost:3306/configuration?serverTimezone=UTC&useSSL=false
+spring.datasource.username=root
+spring.datasource.password=root
+logging.level.root=DEBUG
```
Please don't commit the root logging as debug

jflora91 replied:
> Please don't commit the root logging as debug

Switched to `info` level

ncarapeto commented on configuration-service/src/main/java/tech/plinth/config/interceptor/ConfigureRequestContext.java:
```diff
@@ -0,0 +1,46 @@
+
+        String platformId = httpServletRequest.getHeader(HEADER_PLATFORM_ID);
+
+        if (platformId == null || platformId.isBlank() || platformId.isEmpty()) {
```
replace this line by: `if(Strings.isBlank(platformId)) {`
and by the way, if you test the `platformId.isBlank()` what is the point of testing `platformId.isEmpty()`

jflora91 replied:
> replace this line by: `if(Strings.isBlank(platformId)) {`
> and by the way, if you test the `platformId.isBlank()` what is the point of testing `platformId.isEmpty()`

I understand, `isBlank()` check if the string is empty or contains only white space so there is no need to recheck `isEmpty()`

ncarapeto commented on configuration-service/src/main/java/tech/plinth/config/database/model/Configuration.java:
```diff
@@ -0,0 +1,95 @@
+    @Column(name = "version")
+    private Long version;
+
+    @NotNull(message = "To create a new version is needed data")
```
check grammar

ncarapeto commented on configuration-service/src/main/java/tech/plinth/config/database/model/Configuration.java:
```diff
@@ -0,0 +1,95 @@
+public class Configuration {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO, generator = "native")
```
I'm curious why you chose GenerationType.AUTO? what happens when you have multiple tables with this?
Google a bit about GenerationType.AUTO vs GenerationType.IDENTITY, the other seems more appropriate.

jflora91 replied:
In this case using `GenerationType.AUTO` we can see as a "unique" generator for all tables in the DB (SEQUENCE strategy),  and using `GenerationType.IDENTITY` the id value is set by table.
I have no reason to be using this strategy, just because is the default generation type

ncarapeto commented on configuration-service/src/main/resources/db/migration/V1__create_initial_schema.sql:
```diff
@@ -0,0 +1,7 @@
+create table config (
+    id bigint not null auto_increment,
+    created_at timestamp,
```
why can this one be null? 
check something like: `creation_date timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP`

jflora91 replied:
> why can this one be null?
> check something like: `creation_date timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP`

Can and should be!

ncarapeto commented on configuration-service/src/main/java/tech/plinth/config/database/model/Configuration.java:
```diff
@@ -0,0 +1,95 @@
+        return createdAt;
+    }
+
+    public void setCreatedAt(OffsetDateTime createdAt) {
```
To be honest there is no point for this set. IT should be set automatically instead.
anyway, you are not setting the date anywhere...

jflora91 replied:
> To be honest there is no point for this set. IT should be set automatically instead.
> anyway, you are not setting the date anywhere...

I'm setting automatically as you suggest with `default current_timestamp` in sql and i delete this method.

jflora91 committed:
fix isBlank()/isEmpty() and database query

jflora91 committed:
fix log message

jflora91 committed:
fix pom version definition & Strings.isBlank

jflora91 committed:
review and fix:
message grammar
generationType
update created_at column
clean useless method setCreatedAt

jflora91 committed:
properly format files

jflora91 committed:
clean parent pom

tsbatista commented:
@daveseco7 @ncarapeto please review your open comments, if you are happy with the resolutions, lets approve this PR. For now this seem sane enough to create a baseline for further work

tsbatista commented:
damn the github UI! that close button without confirmation kills me!

daveseco7 commented on configuration-service/src/main/java/tech/plinth/config/database/repository/ConfigurationRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+@Repository
+public interface ConfigurationRepository extends JpaRepository<Configuration, Long> {
+    Configuration findTopByPlatformOrderByVersionDesc(String platformId);
```
You added this query, where are you using it ?

daveseco7 replied:
Make this a optional

daveseco7 commented on configuration-service/src/main/java/tech/plinth/config/delegate/ConfigurationDelegate.java:
```diff
@@ -0,0 +1,66 @@
+        return newVersion;
+    }
+
+
```
Did you understand what the new query does ? If you did this delegate would be way different...

daveseco7 reviewed (commented):
You need to revise your logs. You're pulling too much information which in turn makes the log statments enormous. Log only what you need to debug and understand the problem. Check the logs that are in the request context.

jflora91 commented on configuration-service/src/main/java/tech/plinth/config/delegate/ConfigurationDelegate.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    public Long calculateNextVersionNumber(String platformId) {
+
+        Configuration lastVersion = configurationRepository.findTopByPlatformOrderByVersionDesc(platformId);
```
@daveseco7 
From your previous comments, instead of making the query directly in delegate it's better to separate the business logic creating a service that make the query and send the information to delegate, right?

PS: i'm not finding your comments here

daveseco7 replied:
if you're talking about the first comment, i deleted it because i didn't realise you're already using the query 😄

jflora91 replied:
> if you're talking about the first comment, i deleted it because i didn't realise you're already using the query 😄

lol, yes i'm using but now my doubt is if i'm using in the right place

daveseco7 replied:
Come by my desk tomorrow and we'll have a coffee and you can explain me why it may or may not make sense to have it there 😄

jflora91 committed:
refactor logs and make calculateNewVersionNumber work with Optional

jflora91 committed:
create builder for Configuration
